library/find_unused_disk_module.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/python
+
+DOCUMENTATION = '''
+---
+module: find_unused_disk_module
+short_description: Gets unused disks
+description:
+    - Disks are considered in ascending alphanumeric sorted order.
+    - Disks that meet all conditions are considered 'empty' and returned (using kernel device name) in a list.
+        - 1. No known signatures exist on the disk, with the exception of partition tables.
+        - 2. If there is a partition table on the disk, it contains no partitions.
+        - 3. The disk has no holders to eliminate the possibility of it being a multipath or dmraid member device.
+        - 4. Device can be opened with exclusive access to make sure no other software is using it.
+    - If no disks meet all criteria, "Unable to find unused disk" will be returned.
+    - Number of returned disks defaults to first 10, but can be specified with 'max_return' argument.
+author: Eda Zhou (@edamamez)
+options:
+    option-name: max_return
+    description: Sets the maximum number of unused disks to return.
+    default: 10
+    type: int
+'''
+
+EXAMPLES = '''
+- name: test finding first unused device module
+  hosts: localhost
+  tasks:
+    - name: run module
+      find_unused_disk_module:
+      register: testout
+    - name: dump test output
+      debug:
+        msg: '{{ testout }}'
+'''
+
+RETURN = '''
+disk_name:
+    description: Information about unused disks
+    returned: On success
+    type: complex
+    contains:
+        disks:
+            description: Unused disk(s) that have been found
+            returned: On success
+            type: list
+            samples: ["/dev/sda1", "/dev/dm-0", "/dev/dm-3"]
+                     ["/dev/sda"]
+        none:
+            description: No unused disks were found
+            returned: On success
+            type: string
+            sample: "Unable to find unused disk"
+'''
+
+
+import os
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils import facts
+
+
+def get_disks(run_command):
+    """Return list of devices on the system"""
+    all_disks = run_command(['lsblk', '-o', 'kname', '-l', '-n', '-p'])
```

Comment:
Okay, sorry for this -- I should have noticed it the first time through -- but there are other, better ways to do this.

For one thing, that command will include partitions, which are necessarily not disks.

You could get the exact same list by running `os.listdir('/sys/class/block')`; it's better in that you're using a system call instead of running an executable, but you'll still have the partitions. You could avoid that by instead listing '/sys/block' (Yes, this one contains a subset of the former. Yes, that's a bit goofy.)

Another option that would eliminate the partitions is to use `ansible_facts['devices'].keys()`. It also omits the partitions (it hides them under the disks that contain them). Again, this is better than running an executable since you're just looking at data that's already been gathered.

There might be more work to do here since it's not always trivial to figure out what is and is not a "disk". That can be addressed later, though.

Reply:
I need the complete device paths for some of the other commands (os.open, blkid), but would it be wrong to assume that all the devices will be located in /dev/ ? For example, if `ansible_facts['devices'].keys()` returns sda and dm-0, would it be accurate to assume they are located at /dev/sda and /dev/dm-0?

Reply:
It is safe to assume that they will all be in /dev. The ansible facts get the list of devices by looking at /sys/block, which lists "kernel" devices (always in /dev) --  see [1].


[1] https://github.com/ansible/ansible/blob/0691c9c099d56929321412df1f5842423a2dfdc0/lib/ansible/module_utils/facts/hardware/linux.py#L548